src/net/socket.rs

```diff
@@ -112,71 +138,101 @@ pub struct Socket {
             },
-            len = read_half.read_u32() => Self::read_packet(
-               &mut read_half,
-               len? as usize,
-               &mut output
-            ).await?,
+            Some(message) = stream.next() => {
+               match message {
+                  Ok(Message::Binary(data)) => {
+                     Self::read_packet(data, &mut output).await?
+                  },
+                  Ok(Message::Close(frame)) => {
+                     bus::push(Fatal(anyhow::anyhow!("Relay has been disconnected.")));
+
+                     if let Some(frame) = frame {
+                        log::warn!("Relay has been disconnected, reason: \"{}\", code: {}", frame.reason, frame.code);
+                     } else {
+                        log::warn!("Relay has been disconnected, due to unknown reason.");
+                     }
+
+                     tx.send(Quit)?;
+
+                     break;
+                  },
+                  Err(e) => {
+                     use tokio_tungstenite::tungstenite::{Error::*, error::ProtocolError};
+                     match e {
+                        ConnectionClosed => break,
+                        // Relay can force a closing handshake, and WebSockets requires a closing handshake.
+                        // If we do not get it, it means that relay has been closed and we have to close the session.
+                        AlreadyClosed | Protocol(ProtocolError::ResetWithoutClosingHandshake) => {
+                           log::error!("the connection has been closed, but the client is trying to work with already closed connection. (relay probably crashed)");
+                           bus::push(Fatal(anyhow::anyhow!("Relay has been disconnected.")));
+                           break;
+                        },
+                        _ => anyhow::bail!(e)
+                     }
+                  }
+                  _ => log::info!("got unused message"),
+               }
+            },
             else => (),
          }
       }
       log::info!("receiver loop done");
       Ok(())
    }
 
-   async fn write_packet(
-      write_half: &mut tcp::OwnedWriteHalf,
-      packet: relay::Packet,
-   ) -> anyhow::Result<()> {
+   async fn write_packet(sink: &mut Sink, packet: relay::Packet) -> anyhow::Result<()> {
       let bytes = bincode::serialize(&packet)?;
-      write_half.write_u32(u32::try_from(bytes.len()).context("Packet is too big (wtf)")?).await?;
-      write_half.write_all(&bytes).await?;
+      u32::try_from(bytes.len()).context("Packet is too big (wtf)")?;
+
+      sink.send(Message::Binary(bytes)).await?;
       Ok(())
    }
 
    async fn sender_loop(
-      mut write_half: tcp::OwnedWriteHalf,
+      mut sink: Sink,
       mut input: mpsc::UnboundedReceiver<relay::Packet>,
-      mut quit: oneshot::Receiver<Quit>,
+      quit: (broadcast::Sender<Quit>, broadcast::Receiver<Quit>),
    ) -> anyhow::Result<()> {
+      let (_, mut rx) = quit;
```

Comment:
If you're gonna destructure the tuple, do it in the argument directly or use two separate arguments.
```suggestion
      (quit_tx, quit_rx): (broadcast::Sender<Quit>, broadcast::Receiver<Quit>),
   ) -> anyhow::Result<()> {
```
I don't really see the reason for using a tuple here though so this should be split into two arguments.
```suggestion
      quit_tx: broadcast::Sender<Quit>,
      quit_rx: broadcast::Receiver<Quit>
   ) -> anyhow::Result<()> {
```